### ISSUE TITLE
Undeclared diagnostics.get() reads silently return previous-step data

### ISSUE THREAD

**Issue by duncanwp** [climate-analytics-lab/jax-gcm#673]
Found by a re-audit of the July 2026 ECHAM+RRTMGP review against `dev` @ `f6d1bcd`. Parent: #614.

## What's wrong

`composable_physics.py:247` and `:286` seed each step's diagnostics dict with `{**prev_physics_data}` — the entire previous step's dict is visible to every term from the first term onward. `_validate_ordering` (`:707-722`) checks only *declared* `requires` against upstream `provides`. So any undeclared read silently returns t−1 data.

`requires_audit_test.py` was added since the review and is real hardening: it AST-walks each `PhysicsTerm.__call__` and asserts every bare `diagnostics["key"]` subscript is declared. **But its own docstring exempts the other form:**

> Reads via `diagnostics.get("key")` or behind an `if "key" in diagnostics` guard are optional-by-convention and **exempt**

and `physics_term.py:77-88` *mandates* that optional dependencies be written as `.get()`. **The enforcement mechanism channels every undeclared read into precisely the form nothing checks.** The hole is narrower than before but sharper.

## A live instance, currently shipping

`sundqvist.py:712-717`:
```python
conv = diagnostics.get("convection")
no_convection = (jnp.reshape(conv.ktype, (-1,)) == 0
                 if conv is not None and hasattr(conv, "ktype")
                 else jnp.ones_like(is_ocean, dtype=bool))
```
with the comment at `:702-705` asserting `ktype` comes from the convection diagnostics "when a convection term ran earlier in the step". In the default factory (`echam_terms.py:452-470`) `SundqvistCloudFraction` is index **4** and `TiedtkeConvection` is index **9** — convection runs *later*. The stratocumulus-enhancement gate therefore reads the **previous step's** `ktype` on every step, and the comment states the opposite. Nothing in CI catches it, because the read is a `.get()`.

## The complication a fix has to handle

A blanket "undeclared reads fail" would break the **legitimate** cross-step carry pattern, which uses the same syntax deliberately:
- `surface_physics.py:523` — `diagnostics.get("surface", SurfaceData.zeros(...))`
- `sundqvist.py:745` — `diagnostics.get("clouds", ...)`

The distinction the check needs is **carry-slot vs stale data read**, not declared vs undeclared.

## Fix

Add a debug-mode flag on `ComposablePhysics` (off in production, on in the test suite) that wraps the threaded dict in a small `Mapping` proxy recording which keys came from `prev_physics_data` and have not yet been re-provided this step. On `.get()` of such a key, raise **unless** the key is declared in the reading term's `carry_slots` — that separates the deliberate cross-step contract from an accidental stale read with no false positives. Enable it in `requires_audit_test.py` and the composed-model smoke tests.

Separately, fix the live instance: either move `TiedtkeConvection` before `SundqvistCloudFraction`, or correct the `sundqvist.py:702-705` comment and declare the one-step `ktype` lag explicitly.

## Why it is worth more than it looks

This is not itself a physics error — it is the absence of the only mechanism that would catch a whole class of them. #460 already cost a full investigation to this pattern, and the Sundqvist instance proves the class is not hypothetical.

## Refs

Review finding 3.7. Related: #460 (the regression this class produced before).
